refactor: restructure microagents system (#5886)
All-Hands-AI/OpenHands: microagents/README.md

# OpenHands MicroAgents

> MicroAgents are specialized AI assistants that enhance OpenHands with domain-specific knowledge and task-specific workflows. They help developers by providing expert guidance, automating common tasks, and ensuring consistent practices across projects. Each microagent is designed to excel in a specific area, from Git operations to code review processes.

Reviewer: ```suggestion
MicroAgents are specialized prompts that enhance OpenHands with domain-specific knowledge and task-specific workflows. They help developers by providing expert guidance, automating common tasks, and ensuring consistent practices across projects. Each microagent is designed to excel in a specific area, from Git operations to code review processes.
```

Example repository structure:
```
your-repository/
└── .openhands/
    └── microagents/
        └── repo.md    # Repository-specific instructions
        └── knowledges/  # Private micro-agents that are only available inside this repo
        └── tasks/       # Private micro-agents that are only available inside this repo
```

## Loading Order

When OpenHands works with a repository, it:
1. Loads repository-specific instructions from `.openhands/microagents/repo.md` if present
2. Loads relevant knowledge agents based on keywords in conversations
3. Enable task agent if user select one of them

## Types of MicroAgents

All microagents use markdown files with YAML frontmatter.

### 1. Knowledge Agents

Knowledge agents provide specialized expertise that's triggered by keywords in conversations. They help with:
- Language best practices
- Framework guidelines
- Common patterns
- Tool usage

Key characteristics:
- **Trigger-based**: Activated by specific keywords in conversations
- **Context-aware**: Provide relevant advice based on file types and content
- **Reusable**: Knowledge can be applied across multiple projects
- **Versioned**: Support multiple versions of tools/frameworks

> Example:
> 
> ```markdown
> ---
> name: github
> version: 1.0.0
> agent: CodeActAgent
> triggers:
> - github
> - git
> ---
> 
> You have access to an environment variable, `GITHUB_TOKEN`, which allows you to interact with
> the GitHub API.
> 
> You can use `curl` with the `GITHUB_TOKEN` to interact with GitHub's API.
> ALWAYS use the GitHub API for operations instead of a web browser.
> 
> Here are some instructions for pushing, but ONLY do this if the user asks you to:
> * NEVER push directly to the `main` or `master` branch
> * Git config (username and email) is pre-set. Do not modify.
> * You may already be on a branch starting with `openhands-workspace`. Create a new branch with a better name before pushing.
> * Use the GitHub API to create a pull request, if you haven't already
> * Use the main branch as the base branch, unless the user requests otherwise
> * After opening or updating a pull request, send the user a short message with a link to the pull request.
> * Do all of the above in as few steps as possible. E.g. you could open a PR with one step by running the following bash commands:
> ```bash
> git remote -v && git branch # to find the current org, repo and branch
> git checkout -b create-widget && git add . && git commit -m "Create widget" && git push -u origin create-widget
> curl -X POST "https://api.github.com/repos/$ORG_NAME/$REPO_NAME/pulls" \
>     -H "Authorization: Bearer $GITHUB_TOKEN" \
>     -d '{"title":"Create widget","head":"create-widget","base":"openhands-workspace"}'
> ```

Reviewer: To prevent the documentation from getting too overwhelming, and also to prevent things from getting out-of-date due to the same content being copied in multiple places, maybe we can just make this a link.

```suggestion
You can see an example of a knowledge-based agent in [OpenHands's github microagent](https://github.com/All-Hands-AI/OpenHands/tree/main/microagents/knowledge/github.md).
```

### 2. Repository Agents

Repository agents provide repository-specific knowledge and guidelines. They are:
- Loaded from `.openhands/microagents/repo.md`
- Specific to individual repositories
- Automatically activated for their repository
- Perfect for team practices and project conventions

Key features:
- **Project-specific**: Contains guidelines unique to the repository
- **Team-focused**: Enforces team conventions and practices
- **Always active**: Automatically loaded for the repository
- **Locally maintained**: Updated with the project

> Example structure:
> 
> ```markdown
> ---
> name: repo
> agent: CodeActAgent
> ---
> This repository contains the code for OpenHands, an automated AI software engineer. It has a Python backend
> (in the `openhands` directory) and React frontend (in the `frontend` directory).
> 
> ## General Setup:
> To set up the entire repo, including frontend and backend, run `make build`.
> You don't need to do this unless the user asks you to, or if you're trying to run the entire application.
> 
> Before pushing any changes, you should ensure that any lint errors or simple test errors have been fixed.
> 
> * If you've made changes to the backend, you should run `pre-commit run --all-files --config ./dev_config/python/.pre-commit-config.md`
> * If you've made changes to the frontend, you should run `cd frontend && npm run lint:fix && npm run build ; cd ..`
> 
> If either command fails, it may have automatically fixed some issues. You should fix any issues that weren't automatically fixed,
> then re-run the command to ensure it passes.
> 
> ## Repository Structure
> Backend:
> - Located in the `openhands` directory
> - Testing:
>   - All tests are in `tests/unit/test_*.py`
>   - To test new code, run `poetry run pytest tests/unit/test_xxx.py` where `xxx` is the appropriate file for the current functionality
>   - Write all tests with pytest
> 
> Frontend:
> - Located in the `frontend` directory
> - Prerequisites: A recent version of NodeJS / NPM
> - Setup: Run `npm install` in the frontend directory
> - Testing:
>   - Run tests: `npm run test`
>   - To run specific tests: `npm run test -- -t "TestName"`
> - Building:
>   - Build for production: `npm run build`
> - Environment Variables:
>   - Set in `frontend/.env` or as environment variables
>   - Available variables: VITE_BACKEND_HOST, VITE_USE_TLS, VITE_INSECURE_SKIP_VERIFY, VITE_FRONTEND_PORT
> - Internationalization:
>   - Generate i18n declaration file: `npm run make-i18n`
> ```

Reviewer: ```suggestion
You can see an example of a repo agent in [the agent for the OpenHands repo itself](https://github.com/All-Hands-AI/OpenHands/blob/main/.openhands/microagents/repo.md).
```

### 3. Task Agents

Task agents provide interactive workflows that guide users through common development tasks. They:
- Accept user inputs
- Follow predefined steps
- Adapt to context
- Provide consistent results

Key capabilities:
- **Interactive**: Guide users through complex processes
- **Validating**: Check inputs and conditions
- **Flexible**: Adapt to different scenarios
- **Reproducible**: Ensure consistent outcomes

Example workflow:

> ```markdown
> ---
> name: update_pr_description
> version: 1.0.0
> author: openhands
> agent: CodeActAgent
> inputs:
>   - name: PR_URL
>     description: "URL of the pull request"
>     type: string
>     required: true
>     validation:
>       pattern: "^https://github.com/.+/.+/pull/[0-9]+$"
>   - name: BRANCH_NAME
>     description: "Branch name corresponds to the pull request"
>     type: string
>     required: true
> ---
> 
> Please check the branch "{{ BRANCH_NAME }}" and look at the diff against the main branch. This branch belongs to this PR "{{ PR_URL }}".
> 
> Once you understand the purpose of the diff, please use Github API to read the existing PR description, and update it to be more reflective of the changes we've made when necessary.
> ```

Reviewer: ```suggestion
You can see an example of a task-based agent in [OpenHands's pull request updating microagent](https://github.com/All-Hands-AI/OpenHands/tree/main/microagents/tasks/update_pr_description.md).
```
